otherlibs/runtime_events/runtime_events_consumer.c

```diff
@@ -664,6 +780,294 @@ static int ml_lost_events(int domain_id, void *callback_data, int lost_words) {
   return 1;
 }
 
+
+static value user_events_find_callback_list_for_event_type(value callbacks_root,
+                                                          value event) {
+  CAMLparam2(callbacks_root, event);
+  CAMLlocal2(tmp_callback_array, event_type);
+  tmp_callback_array = Field(callbacks_root, 6); /* ev_user_events */
+  uintnat array_length = caml_array_length(tmp_callback_array);
+
+  uintnat event_index;
+  event_type = Field(event, 2);
+  if (Is_block(event_type)) {
+    event_index = Int_val(Field(Field(event_type, 0), 2));
+  } else {
+    event_index = Int_val(event_type);
+  }
+
+  if (event_index >= array_length) {
+    CAMLreturn(Val_none);
+  }
+
+  CAMLreturn(Field(tmp_callback_array, event_index));
+}
+
+static int user_events_call_callback_list(
+  struct callbacks_exception_holder* holder, value callback_list,
+  value params[4]) {
+  CAMLparam5(callback_list, params[0], params[1], params[2], params[3]);
+  CAMLlocal2(callback, res);
+
+  while (Is_block(callback_list)) {
+      // two indirections as callback is a list item wrapped in a gadt
+    value callback = Field(Field(callback_list, 0), 0);
+    res = caml_callbackN_exn(callback, 4, params);
+
+    if( Is_exception_result(res) ) {
+      res = Extract_exception(res);
+      *holder->exception = res;
+      CAMLdrop;
+      return 0;
+    }
+
+    callback_list = Field(callback_list, 1);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static value caml_runtime_events_user_resolve_cached(
+  value wrapper_root, uintnat event_id, char* event_name,
+  ev_user_ml_type event_type)
+{
+  CAMLparam1(wrapper_root);
+  CAMLlocal3(event, cache_resized, cache);
+
+  cache = Field(wrapper_root, 2);
+
+  if (!Is_block(cache)) {
+    // initialize cache array
+    uintnat new_len = 256;
+    while (event_id >= new_len) {
+      new_len *= 2;
+    }
+
+    cache_resized = caml_alloc(new_len, 0);
+    for (uintnat i = 0; i < new_len; i++) {
+      Field(cache_resized, i) = Val_none;
+    }
+
+    Store_field(wrapper_root, 2, cache_resized);
+    cache = cache_resized;
+  }
```

Comment:
I'm not sure I fully understand the new_len logic, surely we can fix it's size based on `event_id` when it's created for the first time?

Reply:
The logic consisted in having `new_len` to be the smallest power of 2 that is larger than `event_id`, with a minimal value of 256. It's not necessary for the length of the array to be a power of 2, so the logic can be simplified.

Reply:
Oh I see. No, this makes sense and we should keep it given the discussion on doubling elsewhere.